src/main/scala/org/antipathy/mvn_scalafmt/SourcesBuilder.scala

```diff
@@ -0,0 +1,54 @@
+package org.antipathy.mvn_scalafmt
+
+import java.io.File
+import java.util.{List => JList}
+
+import scala.jdk.CollectionConverters._
+
+import org.apache.maven.project.MavenProject
+
+/** Class for building a set of source dirs in the maven project */
+class SourcesBuilder(project: MavenProject) {
+
+  private val sources = Set.newBuilder[File]
+  private val build   = project.getBuild
+
+  private[mvn_scalafmt] def resultSet(): Set[File] =
+    try sources.result()
+    finally sources.clear()
+
+  def result(): JList[File] = resultSet().toList.asJava
+
+  def addMain(dirs: JList[File]): Unit =
+    if (!add(dirs)) {
+      appendPrimary(build.getSourceDirectory)
+      appendAlternative(project.getCompileSourceRoots, build.getOutputDirectory)
+    }
+
+  def addTest(dirs: JList[File]): Unit =
+    if (!add(dirs)) {
+      appendPrimary(build.getTestSourceDirectory)
+      appendAlternative(project.getTestCompileSourceRoots, build.getTestOutputDirectory)
+    }
+
+  private def add(dirs: JList[File]): Boolean = {
+    val ok = dirs != null && !dirs.isEmpty
+    if (ok) sources ++= dirs.asScala
+    ok
+  }
+
+  private def getCanonicalFile(dir: String): File = {
+    val relfile = new File(dir)
+    val absfile = if (relfile.isAbsolute) relfile else new File(project.getBasedir, dir)
```

Comment:
compared to #149: the check for `.isAbsolute` looks to be the primary reason why that commit failed. when testing that build locally, i must not have observed that the list of formatted files was empty (because i didn't expect it).

it took me a while to track this down; I suspected that the move to `maven-core` was the culprit, not this method; neither user who commented on #158 included the full error message (rather than `...`), or it would have been obvious that the project directory had been repeated.